File: app/src/main/java/gardezi/io/amiibofinder/data/AmiiboRepository.java
package gardezi.io.amiibofinder.data;

import android.app.Application;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import gardezi.io.amiibofinder.model.Amiibo;
import io.reactivex.Single;

@Singleton
public class AmiiboRepository {

    private AmiiboWebService webService;

    @Inject
    public AmiiboRepository(AmiiboWebService amiiboWebService) {
        this.webService = amiiboWebService;
    }

    public Single<List<Amiibo>> getAmiibosByName(String name) {
        return webService.getAmiibosByName(name);
    }

}

File: app/src/main/java/gardezi/io/amiibofinder/data/AmiiboRepository.kt
package gardezi.io.amiibofinder.data

import gardezi.io.amiibofinder.data.database.dao.AmiiboDAO
import gardezi.io.amiibofinder.data.network.AmiiboWebService
import gardezi.io.amiibofinder.model.Amiibo
import io.reactivex.Single
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AmiiboRepository @Inject
constructor(private val webService: AmiiboWebService, private val dao: AmiiboDAO) {


    fun getAmiibosByName(name: String): Single<List<Amiibo>> {
        return webService.getAmiibosByName(name).flatMap { amiibos ->
            cacheAmiibos(amiibos)
            Single.just(amiibos)
        }
    }

    private fun cacheAmiibos(data: List<Amiibo>) {
        dao.insertAll(data)
    }

}
